CMakeLists.txt:add custom uninstall target

andy5995 opened the pull request:
So far the only problem I see with this is that when uninstalling, empty directories are not removed.

andy5995 committed:
CMakeLists.txt:add custom uninstall target

klaus-halfmann commented:
Hello @andy5995 welcome to widelands, I compile widelands on macOS with an m1Processor / Apple Silicon.
I am not a big expert with CMake.
* What does this commit provide and why is it needed?
* How can I test this on macOS (currently cloning that repo)

andy5995 commented:
hi @klaus-halfmann .. it provides an 'uninstall' target, so if a 'make install' is run, a 'make uninstall' will remove the files that were installed. (replace 'make' with 'ninja' if you're using that).

When an install is done, an 'install_manifest.txt' file is created in the build directory. when the uninstall target is used, each file from the install manifest is read and is deleted.

More info at https://gitlab.kitware.com/cmake/community/-/wikis/FAQ#can-i-do-make-uninstall-with-cmake

klaus-halfmann commented:
So you add a test for make install / uninstall to the Compile target in gitlhub but only for the  "gcc++-10" compiler.

I could not find the result in https://github.com/widelands/widelands/runs/5963907771?check_suite_focus=true
for (Release, ubuntu-20.04, g++-10) (or https://pipelines.actions.githubusercontent.com/serviceHosts/44802162-6b05-4a72-a3ed-aba8a8afd5e9/_apis/pipelines/1/runs/11056/signedlogcontent/54?urlExpires=2022-04-18T06%3A26%3A13.2494757Z&urlSigningMethod=HMACV1&urlSignature=zJ%2B2KCwbd4rtdb0l2KzxroyfaGrSeInmlsAYh1YMXc4%3D)

But maybe I am missing something?

The macOS way to build the App is found in utils/macos/build_app.sh and triggerd some other way.

I normally use the ./compile.sh script only so please guide me how to call `make install/uninstall`

klaus-halfmann commented:
```
build % ninja install           
[53/58] cd /Users/klaus/develop/widelands/cmake-add-uninstall/build && /opt/homebrew/Cellar/cmake/3.22.3/bin/ctest --output-on-failure
Test project /Users/klaus/develop/widelands/cmake-add-uninstall/build
    Start 1: test_ai
...
100% tests passed, 0 tests failed out of 6
---
[57/58] Install the project...
-- Install configuration: "Debug"
-- Installing: /usr/local/./VERSION  <- Not healthy?
CMake Error at cmake_install.cmake:42 (file):
  file INSTALL cannot copy file
  "/Users/klaus/develop/widelands/cmake-add-uninstall/build/VERSION" to
  "/usr/local/./VERSION": Permission denied.
```

As I am using homebrew (a kind of package manager for macOS) on apple silicon 
my 'usr/local' is actually '/opt/homebrew/' as the other path is reserved for emulated x86 code.

But maybe I cannot  test this on macOS at all?

andy5995 commented:
> So you add a test for make install / uninstall to the Compile target in gitlhub but only for the "gcc++-10" compiler.

That will probably change. It doesn't need to test on all version of all compilers, so I just tried to specify testing on gcc 10.

> 
> I could not find the result in https://github.com/widelands/widelands/runs/5963907771?check_suite_focus=true for (Release, ubuntu-20.04, g++-10) (or https://pipelines.actions.githubusercontent.com/serviceHosts/44802162-6b05-4a72-a3ed-aba8a8afd5e9/_apis/pipelines/1/runs/11056/signedlogcontent/54?urlExpires=2022-04-18T06%3A26%3A13.2494757Z&urlSigningMethod=HMACV1&urlSignature=zJ%2B2KCwbd4rtdb0l2KzxroyfaGrSeInmlsAYh1YMXc4%3D)
> 
> But maybe I am missing something?

I couldn't find it either. I'm not sure what I have wrong so far, so I'm just waiting to have this PR reviewed before I continue. It should still work locally, I just don't have it set up right for testing in the CI.

klaus-halfmann reviewed (approved):
The code looks fine and all the checks passed, so this can go in.
Compiles and starts for me localy.
install does not work on macOS though

No idea however who needs this?

andy5995 commented:
> -- Installing: /usr/local/./VERSION  <- Not healthy?
> CMake Error at cmake_install.cmake:42 (file):
>   file INSTALL cannot copy file
>   "/Users/klaus/develop/widelands/cmake-add-uninstall/build/VERSION" to
>   "/usr/local/./VERSION": Permission denied.
> ```
> 
> As I am using homebrew (a kind of package manager for macOS) on apple silicon my 'usr/local' is actually '/opt/homebrew/' as the other path is reserved for emulated x86 code.
> 
> But maybe I cannot test this on macOS at all?

@klaus-halfmann Looks like you are definitely on the right track. To test installing without root or superuser privileges, you can do

`DESTDIR=$PWD/test_install ninja install`

and afterward, you must specify the same DESTDIR when uninstalling

`DESTDIR=$PWD/test_install ninja uninstall`

'DESTDIR' can be any directory you have write access to.

If using 'make', I think you have to put DESTDIR *after* 'make', i.e. `make DESTDIR=$PWD/test_install install`

andy5995 commented:
> 
> No idea however who needs this?

Anyone who uses `make/ninja install`... That's most developers or people who are cloning the repo, just testing things a bit. But typically as you know, install doesn't need to be used to run the game. But I feel if there's an 'install' target, there should be an 'uninstall' target so one might clean up if he or she uses the install.

klaus-halfmann commented:
```
cd build
DESTDIR=~/develop/widelands/testinsall ninja install
DESTDIR=~/develop/widelands/testinsall ninja uninstall
```
Works but uses a strange structure directly below `testinsall/usr/local/./widelands` but this should be 
checked by some Linux master.

matthiakl reviewed (commented):
(A "Linux master" here :wink: ) 
Instead of DESTDIR there are three possible environment variables to fine tune the installation:
- CMAKE_INSTALL_PREFIX: used for global system files (e.g. icons, binaries). Defaults to /usr/local
- WL_INSTALL_BASEDIR: used for metadata files (version, changelog, license, credits)
- WL_INSTALL_DATADIR: used for game data files

When using these variables during compilation the install and uninstall process works fine as normal user. No permission errors.

For removing empty directories you might try this: https://github.com/xbmc/xbmc/commit/4c8efe4fc7e6bdb759c11a8e1dc6f2422e79bbaa

andy5995 committed:
remove empty directories

Using code by @wsnipex from
https://github.com/xbmc/xbmc/commit/4c8efe4fc7e6bdb759c11a8e1dc6f2422e79bbaa

andy5995 committed:
apply patch by @matthiakl

andy5995 committed:
add test for uninstall

matthiakl reviewed (commented):
This build log output looks suspicious:

>  -- Removing empty dir: {/home/runner/work/widelands/widelands/build/test_install}/usr
-- Removing empty dir: {/home/runner/work/widelands/widelands/build/test_install}
-- Removing empty dir: {/home/runner/work/widelands/widelands/build
-- Removing empty dir: {/home/runner/work/widelands/widelands
-- Removing empty dir: {/home/runner/work/widelands
-- Removing empty dir: {/home/runner/work
-- Removing empty dir: {/home/runner
-- Removing empty dir: {/home
-- Removing empty dir: {

Are these directories removed all the way up to /home? But then the testsuite should fail. Do you get a similar output locally?
Otherwise it looks good now.

andy5995 committed:
fix curly brace placement, add separate section for install/uninstall

andy5995 committed:
remove extra '}'

andy5995 committed:
add '-j$(nproc)' for install/uninstall
